User guide update (#32)
ufs-community/ufs-weather-model: doc/UsersGuide/source/ContributingDevelopment.rst

|zlib 	      | https://github.com/madler/zlib                               | 1.2.11  | $Z_LIB                  |
+-------------+--------------------------------------------------------------+---------+-------------------------+
|libpng       | http://www.libpng.org/pub/png/libpng.html                    | 1.2.44  | $PNG_LIB                |
+-------------+--------------------------------------------------------------+---------+-------------------------+

Compiler:

+--------------------+------------------------------------------------------------------+
|Compiler name       |    Branch name                                                   |
+====================+==================================================================+
| Intel              |   16.0.3.210,16.3.210,18.0.1.163, 18.0.3.222,18.0.5.274, 19.0.2  |
+--------------------+------------------------------------------------------------------+
|impi                |   2018.0.4, 2018.4.274,18.0.1, 2019.2.187                        |
+--------------------+------------------------------------------------------------------+

> These libraries are available on several platforms including Hera, Cheyenne and WCOSS. Installing these libraries on jet and gaea is a work in progress. Users who need to build the libraries. [https://github.com/NOAA-EMC/NCEPLIBS/wiki/Cloning-and-Compiling-NCEPLIBS]

Reviewer: Please revise this so it is a complete sentence

Author: Done

In addition to these libraries, other software is required such as a compiler (Intel or GNU), MPI library (impi), HDF5, cmake, Python, Perl and Ruby to compile the model and run regression tests.

---------------------------------------------
Making code changes using a forking workflow
---------------------------------------------

> If developers would like to make code changes, they need to make a personal fork, set up upstream remote (for merging with the original ufs-weather-model), and create a branch for ufs-weather-model and each of the subcomponent repositories they want to change. They can then make code changes, perform testing and commit the changes to the branch in their personal fork. It is suggested that they merge their branch with the develop branch of the original repositories periodically to get the latest updates and bug fixes.

Reviewer 2: This wording isn't clear - the last sentence - recommend changing "merge their branch", to "update their branch by merging the develop branch..."

Author: Done

5. Checking out the code ufs_release_1.0 should always be as follows:

.. code-block:: console

   git clone https://github.com/ufs-community/ufs-weather-model
   cd ufs-weather-model
   git checkout ufs_release_1.00
   git submodule update --init --recursive

6. Checking out a PR with id ID for testing it should always be as follows:

.. code-block:: console

   git clone https://github.com/ufs-community/ufs-weather-model
   cd ufs-weather-model
   git fetch origin pull/ID/head:BRANCHNAME
   git checkout BRANCHNAME
   git submodule update --init --recursive

It is suggested that the developers inform all the related code managers as the hierarchy structure of the ufs-weather-model repository may require collaboration among the code managers.

-----------------------------------
Engaging in the code review process
-----------------------------------

> When code managers receive a pull request to commit the code changes, it is recommended that they add at least two code reviewers to review the code and at least one of the reviewers has write permission. The reviewers will write comments about the code changes and give a recommendation as to whether the code changes can be committed. What kinds of code changes will be accepted in the repository is beyond the scope of this document; future ufs-weather-model code management may have detailed answer for that.

Reviewer: ```suggestion
When code managers receive a pull request to commit the code changes, it is recommended that they add at least two code reviewers to review the code and at least one of the reviewers has write permission. The reviewers will write comments about the code changes and give a recommendation as to whether the code changes can be committed. What kinds of code changes will be accepted in the repository is beyond the scope of this document; future ufs-weather-model code management documents may have a detailed answer for that.
```

Author: Done

Reviewers may suggest some code changes during the review process. Developers need to respond to these comments in order to get code changes committed. If developers make further changes to their branch, reviewers need to check the code changes again. When both reviewers give recommendation to commit the code, code managers will merge the changes into the repository.

----------------------------
Conducting regression tests
----------------------------

> Only developers that are running on a limited set of platforms (Hera, Cheyenne, WCOSS) can compile and run regression tests using the ufs-weather-model.

Reviewer: ```suggestion
Only developers using Tier 1 and 2 platforms can run the ufs-weather-model regression tests. Other developers need to work with the code managers to assure completion of the regression tests.
```

Author: DOne

To run regression test using rt.sh

rt.sh is a bash shell file to run the RT and has the following options:

.. code-block:: console

   Usage: $0 -c <model> | -f | -s | -l <file> | -m | -r | -e | -h
   -c create new baseline results for <model>
   -f run full suite of regression tests
   -s run standard suite of regression tests
   -l run test specified in <file>
   -m compare against new baseline results
   -r use Rocoto workflow manager
   -e use ecFlow workflow manager
   -h display this help

.. code-block:: console

   % cd ufs-weather-model/tests
   % ./rt.sh -f

> 

Reviewer: Add here

This command can only be used on a NOAA machine using the Intel compiler, where an official baseline is available. For information on testing the CCPP code, or using alternate computational platforms, see the following sections.

This command and all others below produce log output in ./tests/log_machine.compiler. These log files contain information on the location of the run directories that can be used as templates for the user. Each rt*.conf contains one or more compile commands preceding a number of tests.

Author: Done

* What are the requirements for a code to be accepted?

> * What platforms are supported

Reviewer: This introductory bullets can be removed. They were in this file just to remind us of the contents that had to added to the file.

Author: Done

================================
 ufs-community/ufs-weather-model
================================

> 

Reviewer: These lines can be removed since this entire User's Guide is on the ufs-weather-model.

Author: Done

Reviewer: Valbona, thank you for making these changes. When you are ready, please update the files in your fork/branch. This will cause the PR to be updated so we can see the changes. Tks

Author: Ligia, I sent the final copy to Jun and Dusan for review and as soon I have their comments I will update the file in my fork.

> ------------------------------

Reviewer: Before diving into the technical aspects, please include one or more introductory paragraph(s) explaining the big picture of how the community can contribute development.

Author: The ufs-weather-model repository contains the model code and external links needed to build the Unified Forecast System (UFS) atmosphere model and associated components, including the WaveWatch III model. This weather model is used in several of the UFS applications, including the medium-range weather application, the short-range weather application, and the sub-seasonal to seasonal application.

+-------------+--------------------------------------------------------------+---------+-------------------------+
|libpng       | http://www.libpng.org/pub/png/libpng.html                    | 1.2.44  | $PNG_LIB                |
+-------------+--------------------------------------------------------------+---------+-------------------------+

Compiler:

+--------------------+------------------------------------------------------------------+
|Compiler name       |    Branch name                                                   |
+====================+==================================================================+
| Intel              |  16.0.3.210,16.3.210,18.0.1.163, 18.0.3.222,18.0.5.274, 19.0.2   |
+--------------------+------------------------------------------------------------------+
|impi                |  2018.0.4, 2018.4.274,18.0.1, 2019.2.187                         |
+--------------------+------------------------------------------------------------------+

These libraries are available on several platforms including Hera, Cheyenne and WCOSS. Installing these libraries on jet and gaea is a work in progress.

> In addition to these libraries, other software is required such as a compiler (Intel or GNU), MPI library (impi), HDF5, cmake, Python, Perl and Ruby to compile the model and run regression tests.

Reviewer: Please direct the reader to documentation of the NCEPLibs superproject in case they need to build the libraries. [https://github.com/NOAA-EMC/NCEPLIBS/wiki/Cloning-and-Compiling-NCEPLIBS]

Author: Done

      # Create an issue in the authoritative repository. For example to commit code changes to fv3atm, please go to https://github.com/NOAA-EMC/fv3atm, under NOAA-EMC/fv3atm and find the “Issues” tab next to the “Code” tab. Click on “Issues” and a new page will appear. On the right side of the page, there is a green “New issue” button. Clicking on that will lead to a new issue page. Fill out the title, comments to describe the code changes, and also please provide personal fork and branch information. Lastly, click on the “Submit new issue” button, so that the new issue is created.

      # When the development is mature, tests have been conducted, and the developer is satisfied with the results, create a pull request to commit the code changes.

      * Merge developer’s branch to the latest ufs-weather-model develop branch in authoritative repository. If changes are made in model sub-components, developers need to merge their branches to branches with the corresponding authoritative repository (or original repository for some components). For this, code management practices of the subcomponents need to be followed.

>       * Regression tests need to pass on at least one supported platform.

Reviewer: ```suggestion
      * Regression tests associated with the ufs-weather-model are available in Tier 1 and Tier platforms as described in https://github.com/ufs-community/ufs-weather-model/wiki/Weather-Model-Platform-and-Compiler-Support. If the developer has access to these platforms, the developer should pass the regression test on at least one supported platform. If the developer does not have access to these platforms, this should be stated in the PR so the code manager(s) can conduct the tests. 
```

Author: Done

>       * For each component branch where developers make changes, developers need to go to their personal fork on GitHub and click on the “New pull request” button. When a new page “Compare changes” appears, developers will choose the branch in their fork with code changes to commit and the branch in upstream repository that the changes will be committed to. The code differences between the two branches will be displayed. Developers can review the differences and click on “submit pull request” to make the pull request. After code changes are committed to the component repository, developers will make pull requests to ufs-weather-model repository.

Reviewer: Please consider including info from Dom and other code managers on procedure for nested PRs in the UFS. Rewrite as needed for clarity.

(1) When PRs are created, the creator must temporarily modify .gitmodules to point to his/her fork and branch if updates are required for submodules.

(2) Merging code from PRs with submodules requires coordination with the person making the PRs. From the "innermost" nested PR up to the top-level PR, the PRs need to be merged as-is. After each merge, the person creating the PRs has to update his/her local code to check out the merged version, revert the change to .gitmodules, and push this to GitHub to update the PR. And so on and so forth.

(3) Checking out the code ufs_release_1.0 should always be as follows:
git clone https://github.com/NOAA-EMC/NCEPlibs
cd NCEPlibs
git checkout ufs_release_1.00
git submodule update --init --recursive

(4) Checking out a PR with id ID for testing it should always be as follows:
git clone https://github.com/NOAA-EMC/NCEPlibs
cd NCEPlibs
git fetch origin pull/ID/head:BRANCHNAME
git checkout BRANCHNAME
git submodule update --init --recursive

For my current NCEPLIBS PRs, this is simple: first, accept and merge the cmake PR - which you did. Then let me update cmake to point back to NOAA-EMC in all repositories that use the cmake modules repository as submodule - which I did. Then merge the NCEPLIBS-xyz PRs (can be done in parallel) - hasn't been done yet. Then let me update my NCEPLIBS umbrella PR to point the submodules back to NOAA-EMC. Then merge this one. Done.

Author: Done

> It is suggested that the developers inform all the related code managers as the hierarchy structure of the ufs-weather-model repository may require collaboration among the code managers.

Reviewer: Please state how the developer can find out who the code manager(s) is/are.

Reviewer: How can the developer find out who the code managers are so they can inform them?

-----------------------------------
Engaging in the code review process
-----------------------------------

When code managers receive a pull request to commit the code changes, it is recommended that they add at least two code reviewers to review the code. The reviewers will write comments about the code changes and give a recommendation as to whether the code changes can be committed. What kinds of code changes will be accepted in the repository is beyond the scope of this document; future ufs-weather-model code management may have detailed answer for that.

Reviewers may suggest some code changes during the review process. Developers need to respond to these comments in order to get code changes committed. If developers make further changes to their branch, reviewers need to check the code changes again. When both reviewers give recommendation to commit the code, code managers will merge the changes into the repository.

---------------------------
Conducting regression tests
---------------------------

> Only developers that are running on a limited set of platforms (Hera, Cheyenne, WCOSS) can compile and run regression tests using the ufs-weather-model.

Reviewer: Suggest changing to: Regression tests for the ufs-weather-model can only be run by developers with access to Tier 1 and Tier 2 platforms.

To run regression test using rt.sh

.. code-block:: console

   % cd ufs-weather-model/tests
   % ./rt.sh -f

> 

Reviewer: Include introduction to rt.sh. Pull content from Sections 9.2.3.1 of the CCPP TechDoc (https://ccpp-techdoc.readthedocs.io/en/latest/BuildingRunningHostModels.html#running-the-ufs-atmosphere-using-the-regression-tests-rts). Source code is from https://github.com/NCAR/ccpp-doc/blob/master/CCPPtechnical/source/BuildingRunningHostModels.rst

Regression test log files (ufs-weather-model/tests/Compile_$(MACHINE_ID).log and ufs-weather-model/tests/RegressionTests_$(MACHINE_ID).log ) will be updated.

> 

Reviewer: Provide explanation for when/why developer needs to create new baseline. Use content from Section 9.2.3.2 of CCPP TechDoc.

Reviewer: 9.2.3.2. Baselines
Regression testing is only possible on machines for which baselines exist, that is Tier 1 and 2 platforms.

When porting the code to a new machine, it is useful to start by establishing a personal baseline. Future runs of the RT can then be compared against the personal baseline to ascertain that the results have not been inadvertently affected by code developments. The rt.sh -c option is used to create a personal baseline.

./rt.sh -l rt.conf -c fv3 # create own reg. test baseline
Once the personal baseline has been created, future runs of the RT should be compared against the personal baseline using the -m option.

./rt.sh -l rt.conf -m # compare against own baseline

To create new baseline:

.. code-block:: console

   % cd ufs-weather-model/tests

>    % ./rt.sh -f -c

Reviewer: Somewhere include list of command line options for rt.sh. Pull content from CCPP TechDoc Section 9.2.3.3

Author: Done

      * To run regression test using NEMSCompsetRun

.. code-block:: console

   % cd ufs-weather-model
   % ./NEMS/NEMSCompsetRun -f 

Regression test log files (ufs-weather-model/log/$MACHINE_ID/* ) will be updated.

To create new baseline:

.. code-block:: console

   % cd ufs-weather-model
   % ./NEMS/NEMSCompsetRun--baseline fv3 --platform=${PLATFORM}

The value of ${PLATFORM} can be found in ufs-weather-model/compsets/platforms.input.

Developers need to commit the regression test log files to their branch before making pull request.

-------------------------------------
Compiling the code and running a test
-------------------------------------

> Currently developers running on Hera, Cheyenne or WCOSS can compile and run tests using the ufs-weather-model. In the document below, ufs-weather-model directory points to a branch in the developer’s personal fork.

Reviewer: Instead of listing specific platforms, just say Tier 1 and Tier 2 machines and point reader to platform policy in wiki. That way this documentation will not age off as more machines are added.
https://github.com/ufs-community/ufs-weather-model/wiki/Weather-Model-Platform-and-Compiler-Support

      * Compile and run a test using rt.sh

	i) compile the code

Developers can compile the ufs-weather-model using the compile script ufs_weather_model/tests/compile.sh. To compile the code:

.. code-block:: console

   % git clone --recursive https://github.com/ufs-community/ufs-weather-model
   % cd ufs-weather-model/tests
   % ./compile.sh PATHTR MACHINE_ID MAKE_OPT BUILD_NR

Where:

.. code-block:: console

   PATHTR: the full path of FV3 directory under ufs-weather-model
   MACHINE_ID: machine ID, e.g: wcoss_cray, hera.intel
   MAKE_OPT: compile options, default (‘’, empty string) is for 64-bit OpenMP non-hydrostatic build using AVX2, other options are:
   -       ‘DEBUG=Y’: turn on debug option
   -       ‘VERBOSE=Y’: turn on VERBOSE mode to get additional details on compile
   -       ‘OPENMP=Y’: use openmp
   -       ‘AVX2=Y’: use AVX2 in Intel Haswell for better performance.
   -       ‘HYDRO=Y’: hydrostatic mode
   -       ‘CCPP=Y’: using ccpp framework and physics
   -       ‘STATIC=Y’: when CCPP=Y, using static build instead of dynamic build

>    -       ‘SUITE=xxx': CCPP physics suite, e.g. “FV3_GFS_2017_gfdlmp”

Reviewer: ```suggestion
   -       ‘SUITES=xxx, yyy, zzz': CCPP physics suites, e.g. “FV3_GFS_2017_gfdlmp”
```

Author: Done

   BUILD_NR: the number of build (there might be several copies of the executable). The final executable would be fv3_${BUILD_NR}.exe

Example:

.. code-block:: console

   %  ./compile.sh /gpfs/hps/emc/global/noscrub/First.Last/ufs-weather-model/FV3 wcoss_cray '' 1

> The executable generated from the compile.sh can be used in global workflow to run experiments.

Reviewer: ```suggestion
The executable generated from the compile.sh can be used in workflows that do not use CIME (Community Infrastructure to Model the Earth) to run experiments.
```

Author: Done
